### ISSUE TITLE
RGGB compression somtimes worse than RawSpeedCompress+LZMA2

### ISSUE THREAD

**Issue by StephanBusch** [FLIF-hub/FLIF#151]
In my ongoing tests with RGGB created by dcraw compression of FLIF is sometimes worse
than RawSpeedCompress+LZMA2. RawSpeedCompress uses different Delta versions for uncompressed pixel data and applies ZigZag-encoding which helps all kinds of general purpose compressors.

https://drive.google.com/file/d/0ByLIAFlgldSoNFR6RmcwdTgxQTQ/view?usp=sharing
leica_m82_05.rggb 11.359.612 (FLIF 0.1)
leica_m82_05.rggb 10.928.429 (FLIF 0.1.3_slower_but_stronger -n)
leica_m82_05.rggb   5.560.365 (RawSpeedCompress3 + LZMA2)

https://drive.google.com/file/d/0ByLIAFlgldSoalhkSkdsNmJNaDA/view?usp=sharing
nikon_1_v2_17.rggb 9.931.720 (FLIF 0.1)
nikon_1_v2_17.rggb 9.636.696 (FLIF 0.1.3_slower_but_stronger -n)
nikon_1_v2_17.rggb 8.810.004 (RawSpeedCompress3 + LZMA2)

https://drive.google.com/file/d/0ByLIAFlgldSoZmJDUXl2YWVPS2M/view?usp=sharing
nikon_d5200_14.rggb 24.334.053 (FLIF 0.1)
nikon_d5200_14.rggb 23.684.359 (FLIF 0.1.3_slower_but_stronger -n)
nikon_d5200_14.rggb 21.527.675 (RawSpeedCompress3 + LZMA2)

Can you please check why FLIF is worse here?
Will future versions also get something like Delta+ZigZag+LZ?

In almost every case the -n switch provided better compression on RGGB

**Comment by psykauze** [comment]
Could you try renaming the RGGB file in PPM and compression it in FLIF interlaced ?

**Comment by matthiaskrgr** [comment]
I wrote a simple crusher script for flif (matthiaskrgr/flifcrush), you can try running it on one of the images:
````
export FLIF=/path/to/flif/binary
main.py nikon_d5200_14.rggb
````
It will take a lot of time but it might get the size down a bit further.

**Comment by jonsneyers** [comment]
I found out what is happening. I know how to solve this. I just need to implement it. Soon FLIF will beat that other method.

**Comment by jonsneyers** [comment]
Here you go:
```
11,359,333  leica_m82_05.rggb.flif0.1.3
5,117,762  leica_m82_05.rggb.flif0.1.4
8,450,275  nikon_1_v2_17.rggb.flif0.1.4
2,0384,559 nikon_d5200_14.rggb.flif0.1.4

**Comment by StephanBusch** [comment]
There are still RGGB that compress better with Delta + ZigZag + LZMA2 than they do with FLIF:

https://drive.google.com/file/d/0ByLIAFlgldSoNlR4ZURaTUE2QTg/view?usp=sharing
pentax_q10_19.rggb 10.692.568 (FLIF 0.1.4)
pentax_q10_19.rggb 10.167.315 (Delta + ZigZag + LZMA2)

https://drive.google.com/file/d/0ByLIAFlgldSoSkZyQzJvWFNpQTQ/view?usp=sharing
panasonic_lumix_dmc_gh3_10.rggb 14.186.367 (FLIF 0.1.4)
panasonic_lumix_dmc_gh3_10.rggb 13.496.039 (Delta + ZigZag + LZMA2)

PLC helped a lot - on most files FLIF is stronger than without it, and sometimes it also
outperforms RawSpeedCompress3 + Delta + ZigZag + MCM 0.83 -x9
but I think it has the capability of beating it on every RGGB image.
can you check that again, please?
See the almost complete test here: http://www.squeezechart.com/camera-raw.html

**Comment by psykauze** [comment]
As I see, Flif outperform the other method by >1% on average, that is really good. Because each compressor has its own method, sometimes one compressor that usually be worse than others can be more efficient for some case.

You can check some plots at issue #144 you will see that FLIF can sometimes outperform JPEG in its own sandbox, and sometimes webp outperform FLIF.

**Comment by jonsneyers** [comment]
The Bayer CFA layout is different for different camera models; FLIF currently assumes some specific subpixel layout for its YIQ transform, and if this assumption is wrong, then compression suffers a bit.
You can use `flif -R` to disable YIQ. Also, disabling channel compactification with `-C` can help.

14,005,394 panasonic_lumix_dmc_gh3_10.rggb.R.flif
13,416,647 panasonic_lumix_dmc_gh3_10.rggb.RC.flif

10,481,645 pentax_q10_19.rggb.R.flif
10,481,603 pentax_q10_19.rggb.RC.flif

Does anyone have a good source of information on these subpixel layouts? I tried reading the `dcraw` source code, but that code is not very readable...

**Comment by psykauze** [comment]
Dcraw's exif output can show the CFA pattern as a matrix (Color1, Color2)(Color3, Color4). DNGs and many (all?) RAWs formats has an Exif data for it.

**Comment by StephanBusch** [comment]
maybe RawSpeed can help?
https://github.com/klauspost/rawspeed/tree/Rawspeed
they have a camera.xml database for supporting new cameras:
https://github.com/klauspost/rawspeed/blob/develop/cameras-xml.md

**Comment by psykauze** [comment]
I found that Exiftools does not mention the CFA pattern of the RW2, libopenraw documentation say that the pattern should be BGGR and DCRAW say the CFA pattern of the file is GBRG.

Which one to trust ?

**Comment by jonsneyers** [comment]
Also: how does DCRAW output the extracted data?
When FLIF reads a .rggb file, it is using the following layout, which I _think_ is what DCRAW produces:

GG
RB

Another complication is that these camera raw formats do not seem to have the same range for R, G and B. I'm not even sure if G1 and G2 are always in the same range. So that also messes up the YIQ.

Probably the way to get best compression (if you're not interested in progressive decoding) is to use `flif -Rbn`, with or without `-C`. You can usually improve further by using higher `-r` and `-S` (simultaneously), e.g. `flif -RCbnS120 -r8`.

**Comment by psykauze** [comment]
I'm sorry but DCRAW outputs with the original pattern, the most common case is:
RG
GB

I'm working at image-rggb.cpp there is also an issue for maxval < 0xff at blue plane

**Comment by jonsneyers** [comment]
Thanks, @psykauze . Now we should probably preprocess raw input to be in RG GB format.

**Comment by psykauze** [comment]
You should also change the decoder part ;)

**Comment by psykauze** [comment]
So...
I've made some analysis on the RAW files and found there is also rotation to take account. The results after rotating images for getting the right CFA pattern is worse than before.

Here the new results:
[log-newRGGB.txt](https://github.com/FLIF-hub/FLIF/files/34464/log-newRGGB.txt)

**Comment by jonsneyers** [comment]
Maybe one image per camera model is not enough to really draw conclusions, but I wonder what is going on here. The differences in compression when changing the CFA pattern are not huge anyway, which suggests to me that doing the YIQ transform on R G1 B while storing 1+G2 in the 'alpha' channel is not the best approach -- it was just a quick & dirty hack anyway, to match RGGB with the RGBA functionality that was already there. Maybe a custom color transform for RGGB would be better.

**Comment by StephanBusch** [comment]
I would also vote for a custom color transformfor RGGB.
At least for Sony there is an explanation:
http://diglloyd.com/blog/2014/20140212_2-SonyA7-RawDigger-posterization.html

The Fuji CFA pattern is different from others - see wikipedia:
https://en.wikipedia.org/wiki/Bayer_filter

**Comment by psykauze** [comment]
That's why I disqualified the fuji x e1.

Also, I would like to try the R, (G1+G2)/2, B, (G1-G2) algo.

You're right too concerning the R, G and B ranges, there's is correction factor for each color.

**Comment by jonsneyers** [comment]
Converting to R, (G1+G2)/2, B, (G1-G2) is not lossless: one bit is lost in the division by two and it cannot be recovered. But we could try to directly apply a modified variant of YIQ, maybe something like this:
Y = R+B+G1+G2
I = R - B
Q = R+B -G1-G2
Gdiff = G1-G2

I'm not sure if encoding G1-G2 is really more efficient than just encoding G1 or G2. The difference is expected to be closer to zero, but it needs an extra bit to be represented.

**Comment by psykauze** [comment]
There's no lost on R, (G1+G2)/2, B, (G1-G2) conversion if you check if numbers is odd or even:

X = (G1+G2)/2 (Always round to lower int like N.5 => N); W = G1-G2;
Case Y odd: G1 = X + W/2 ; G2 = X - W/2 ;
Case Y even: G1 = X + (W+1)/2 ; G2 = X - (W-1)/2;

You're right, difference need to be represented with an extra bit but I think this W extra bit can be moved to X in some cases (like if Xdec+Wdec/2 > 0xFFFF then that means Xdec = Xenc << 1 + extra_bit). I need some calculations to check if it is possible.

**Comment by jonsneyers** [comment]
What is Y in that case distinction?

**Comment by psykauze** [comment]
Y was W I've edited the post sorry.

**Comment by jonsneyers** [comment]
Ah, right, the parity of the sum is equal to the parity of the difference so there would indeed be enough information to restore the exact sum. I'll try that transformation.

Also I'm assuming that raw files have no more than 14 bpc, is that a valid assumption? It helps if we want to represent G1-G2 in an uint16_t...

**Comment by psykauze** [comment]
I haven't see yet sensors up to 14bits so I think this is a valid assumption.

Be carreful, (G1 - G2) is a signed value but maybe you want store the data as '(Diff << 1) + sign' ?

**Comment by jonsneyers** [comment]
I'm storing it as 0x4000 + G1 - G2, which is always > 0 if the numbers are at most 0x3fff.

But unfortunately, my preliminary results (just testing on a single image) seem to indicate that any manipulation on the R G1 G2 B values (sum, avg, diff) just makes compression worse.

**Comment by psykauze** [comment]
Could you send me your new image-rggb.cpp, please ? I would like to try some things with the Stephan set.

**Comment by jonsneyers** [comment]
I pushed it as commented-out code (only in the rggb loading part, not in the decode/save part),

**Comment by psykauze** [comment]
I've tried with R,mean(G),B,diff(G) with many hacks but it is still very worse (+20-50%) than before.

Also, I think I understand why the results was worse after we corrected the RGGB pattern. The red plane is about half the luminosity than it should (usualy corrected by RAW decoders) and blue plane is ~70%. So maybe the YIQ transform is not efficient in this case.

I will try this hack: R => Alpha plane, G2 => Red Plane, G1 => Green plane, B => Blue Plane and compare results.

**Comment by StephanBusch** [comment]
There are also Bayer sensors that capture in 16-Bit.
For example Mamiya-Leaf and Phase One raw images and also Hasselblad raw images.
I will upload some samples (converted with DCRAW) later.

**Comment by psykauze** [comment]
I made some (long) tests in interlaced mode with some hacks:
1) I've rotated and mirrored the greyscales for having the same (RGGB) pattern on every images. You should see the transformation in the filename.
2) I've disqualified the sigma and the finepix x-e1 pictures due to their attypic CFA pattern
3) Test named "before" is before modifications of image-rggb.cpp
3) Test "RGGB corrected" is just after finding there is an issue to the demosaicing algo in image-rggb.cpp
3) Test "RGGB+bpp": I've made a bit-depth detection before initialising the planes
4) Test "RGGB+invertG2": I've only inverted the alpha plane (representing G2) with "maxval - pixel" instead of "1 +pixel"
5) Test "bpp+invertG2": All corrections above
6) Tests "G2G1BR", "G2G1RB", "G2RBG1" and"G1RBG2": I've tried to change planes order to see what happening (G2RBG1 is like before the RGGB correction)
7) Test "all invert": All planes was inverted ("maxval - pixel")

[image-rggb-patch.txt](https://github.com/FLIF-hub/FLIF/files/34772/image-rggb-patch.txt)
[results-psykauze.txt](https://github.com/FLIF-hub/FLIF/files/34773/results-psykauze.txt)

**Comment by jonsneyers** [comment]
I added these results to the spreadsheet over here:
https://docs.google.com/spreadsheets/d/1LxY78fbm47VmrYGTXkBXXitGjhGl32NsuHPH2QXufgA/edit?usp=sharing
I made the results a bit more visual using color.

The order of the RGGB planes makes surprisingly little difference (at most 3%, and typically less than 1%).

Is maxval-pixel safe? If the alpha plane value is zero, the other planes don't get encoded...

**Comment by psykauze** [comment]
maxval-pixel is as safe than 1+pixel, if "pixel=maxval" you will get a fully-transparent pixel.

Also, my bit-depth detection can fail if the maxval measured = 2^N

I've edited the bit-depth detection, so in case it measure maxval = 2^N it make the bit-depth 2^(N+1)

**Comment by psykauze** [comment]
I think I understand why the plane ordering change the compression ratio.

R, Gs and B planes have not the same luminosity factor. Also G1 and G2 are just slighty different so YIQ transform with this 2 planes should be more efficient.

In natural environment, Red plane is more different of green planes than blue of green.

(For info: Red coefficient on RAW is 0,5, Green = 1, Blue = 0,7)

**Comment by StephanBusch** [comment]
16-bit RGGB

Phase One:
https://drive.google.com/file/d/0ByLIAFlgldSoU2dyOXptOUxxdjg/view?usp=sharing
Mamiya-Leaf:
https://drive.google.com/file/d/0ByLIAFlgldSoTDB6b0NNOWxjWjQ/view?usp=sharing
Hasselblad:
https://drive.google.com/file/d/0ByLIAFlgldSoTk96NU4xTmlrTVU/view?usp=sharing

**Comment by jonsneyers** [comment]
The Phase One image seems to use 15 bits on R and B and 16 bits on G1 and G2.
The Mamiya-Leaf image seems to use 13 bits on R and B and 14 bits on the greens.
The Hasselblad image seems to use 15 bits on all colors.
(I'm just looking at the number of distinct values per channel; the next power of two is what I guess the bit depth is)

So for now, putting 1+R at alpha is safe ;) but of course it would be cleaner to actually distinguish RGGB from RGBA. Probably best to add an extra template argument to the encoder and decoder to disable the A=0 branches, that would also improve the speed of encoding/decoding flat RGB images.

**Comment by psykauze** [comment]
I've tried to (hardly) deactivate the alpha-cleanup but "some" files fails in decoding. When the filesize is not the same than "before" that means the flif file is broken.

**Comment by jonsneyers** [comment]
Deactivating alpha=0 interpolation alone is not enough. There are also changes inside the inner loops. I'll try take a look at it tomorrow. Added a new issue for this, since it is also useful for RGB images (some small speedup to be expected) and RGBA images where you for some reason want to store the original RGB values at A=0.

**Comment by psykauze** [comment]
@StephanBusch The "Fujifilm x e1" CFA is [GRBGBR BGGRGG RGGBGG GBRGRB RGGBGG BGGRGG] I've tried to make a new preprocessor for this kind of CFA pattern but it could be difficult. Also because there is 20 Green pixels, 8 Blue pixels and 8 Red Pixels. So, for now you should compress it in greyscale mode (*.pgm) instead of (*.rggb).

Also, there is not yet an "Orientation" detection. FLIF assume that the left-up corner pixel is a red pixel, so you have to rotate the file manualy.

**Comment by jonsneyers** [comment]
Here are some results for those 3 huge test images @StephanBusch provided (thanks for the test images, btw!)

```
-rw-r----- 1 jon jon  80300323 Nov 15 21:44 20130624-CF043185.pgm
-rw-r--r-- 1 jon jon  45620822 Nov 15 23:35 20130624-CF043185.pgm.flif
-rw-r--r-- 1 jon jon  65960722 Nov 15 23:42 20130624-CF043185.pgm.png
lrwxrwxrwx 1 jon jon        21 Nov 15 23:56 20130624-CF043185.pgm.rggb -> 20130624-CF043185.pgm
-rw-r--r-- 1 jon jon  43346608 Nov 15 22:28 20130624-CF043185.pgm.rggb.flif
-rw-r--r-- 1 jon jon  43248365 Nov 16 00:14 20130624-CF043185.pgm.rggb.new.flif
-rw-r----- 1 jon jon 121768435 Nov 15 21:50 Hasselblad_H4D-60.pgm
-rw-r--r-- 1 jon jon  67266732 Nov 15 23:48 Hasselblad_H4D-60.pgm.flif
-rw-r--r-- 1 jon jon  98908643 Nov 15 23:49 Hasselblad_H4D-60.pgm.png
lrwxrwxrwx 1 jon jon        21 Nov 15 23:56 Hasselblad_H4D-60.pgm.rggb -> Hasselblad_H4D-60.pgm
-rw-r--r-- 1 jon jon  66591477 Nov 15 22:36 Hasselblad_H4D-60.pgm.rggb.flif
-rw-r----- 1 jon jon 112008019 Nov 15 21:50 Wales_4_10_91.pgm
-rw-r--r-- 1 jon jon  60465878 Nov 15 23:53 Wales_4_10_91.pgm.flif
-rw-r--r-- 1 jon jon  86270819 Nov 15 23:50 Wales_4_10_91.pgm.png
lrwxrwxrwx 1 jon jon        17 Nov 15 23:56 Wales_4_10_91.pgm.rggb -> Wales_4_10_91.pgm
-rw-r--r-- 1 jon jon  58670752 Nov 15 22:35 Wales_4_10_91.pgm.rggb.flif
-rw-r--r-- 1 jon jon  58609200 Nov 16 00:18 Wales_4_10_91.pgm.rggb.new.flif

**Comment by psykauze** [comment]
It seems the RGGB preprocessor is not very efficient with the Hasselblad image. Is the CFA pattern "RGGB" ? Is the top-left pixel Red ?

**Comment by jonsneyers** [comment]
No idea. @StephanBusch do you have the metadata from the original raw files? Also, I would be interested to know what format those raw files used originally, and how big those files were.

**Comment by StephanBusch** [comment]
here are the original raw files:

Hasselblad:
https://drive.google.com/file/d/0ByLIAFlgldSoUlUxWGtFVGJXTVE/view?usp=sharing
Mamiya-Leaf:
https://drive.google.com/file/d/0ByLIAFlgldSoaEJxeUE0WlVXQTA/view?usp=sharing
Phase One:
https://drive.google.com/file/d/0ByLIAFlgldSoUGxJaDJzU1BtRmc/view?usp=sharing

I can upload more Raw if you want
Thank you for the info @JonSneyers and @psykauze

Can the compression of Sony raw also be improved - given the info about 11bit +7 bit architecture?
http://diglloyd.com/blog/2014/20140212_2-SonyA7-RawDigger-posterization.html
This compression scheme is used on all Sony raw by default

**Comment by StephanBusch** [comment]
the Sigma Raw do not use a color filter array but the Foveon X3 sensor. Should they also be compressed as pgm?

**Comment by psykauze** [comment]
For this kind of sensor (Foveon X3), no need to preprocess the data because each pixel is directly encoded as a RGB value. You should encode it as a ppm (P6). Encode it with other methods will not improve compression.

**Comment by psykauze** [comment]
So, after raw analysis, the Hasselblad and the Phaseone files are rotated by "270 CW" so they should rotated by 90° (convert -rotate 90 $1 $2) more for matching the right CFA Pattern.

The Mamiya file is horizontaly oriented so no need to change orientation.

**Comment by psykauze** [comment]
After learning how the YIQ transform is made, I've made some new test with G1BG2-R ordering planes and as I expected compression improve.
G1BG2-R method beat all the time the RG1B-G2 method (with this set)

Using R instead of B could be better in some case (peoples blue wears/jeans, blue signs,..?)
Switching G1 with G2 should not significantly improve the compression.

![results-with-g1bg2r](https://cloud.githubusercontent.com/assets/7633907/11221741/fad9564a-8d66-11e5-884c-37b6d72f628e.png)

**Comment by jonsneyers** [comment]
You can forget the --keep-alpha-zero switch, if the input file is .rggb it is set automatically :)

**Comment by psykauze** [comment]
Thank you for the info :)

Also, I made tests for @StephanBusch with both interlaced and non-interlaced coding

[FLIF_0.1.6_G1BG2-R.csv.txt](https://github.com/FLIF-hub/FLIF/files/37262/FLIF_0.1.6_G1BG2-R.csv.txt)

**Comment by StephanBusch** [comment]
thank you @psykauze

**Comment by StephanBusch** [comment]
I get different results with FLIF 0.1.6 so I guess that G1BG2-R ordering planes you did is not yet implemented

**Comment by psykauze** [comment]
Indeed, because it's just a minor correction. This evening I will work on more improvements, I will push this correction after or maybe @jonsneyers will do it.

RGGB method seems (to me) not implemented in the best way. For now, it's just an hack and users should manipulate image for matching the RGGB CFA (-rotate -flip -flop...) and FLIF can't handle (yet) the image orientation that is an issue.

Could you send me your results please ?

**Comment by StephanBusch** [comment]
here are my results.
None of the images were rotated.

[camera-raw.xlsx](https://github.com/FLIF-hub/FLIF/files/38016/camera-raw.xlsx)

**Comment by StephanBusch** [comment]
it would be cool if FLIF would rotate the RGGB automatically or it tries all 4 orientations and picks the best - may rotation also help on RGB images?

**Comment by psykauze** [comment]
@StephanBusch "may rotation also help on RGB images?" -> No, rotation "should" not change anything also for RGGB method.

FLIF has a bug that it can't detect the CFA Pattern "yet" so for now it only work with a [R G1][G2 B] RGGB file that's why we need to rotate the RGGB file for matching this CFA Pattern.

I've made raw2rggb script that detect "some" CFA patterns and rotate the rggb to the right way. (But this is not working with Windows, sorry)

**Comment by psykauze** [comment]
I've tested raw2flif with the RAW files of the usual set. Here the result.
[raw2flif.csv.txt](https://github.com/FLIF-hub/FLIF/files/40863/raw2flif.csv.txt)

raw2flif fail with the Mayima-Leaf and the Phase One raws because dcraw didn't read they.
Also, raw2flif fail with the fujifilm x e1 and the sigmas raw because of their attypic sensors.

**Comment by psykauze** [comment]
There is an issue with raw2rggb and Phase_One_20130624-CF043185.iiq. I don't know why but the CFA pattern is not correctly decoded by dcraw(???). So, my assumption and the raw2rggb assumption is wrong. I've manualy converted it and I got ~1% compression improvement (~41,0MiB unless 41,4MiB).

**Comment by StephanBusch** [comment]
I cannot use that scripts raw2flif and raw2rggb because I don't have a linux system.
Here are the results of FLIF 0.1.7 -n without scripts and without rotation:
http://www.squeezechart.com/camera-raw4.html

**Comment by psykauze** [comment]
@StephanBusch Here is a *.bat that do the same as raw2rggb (It's buggy but it works) You need exiftool.exe, dcraw.exe and sed.exe (and its dependencies, see GnuWin website) in the same directory.
[raw2rggb.bat.txt](https://github.com/FLIF-hub/FLIF/files/43659/raw2rggb.bat.txt)

@jonsneyers Please add this script in the root directory please. Also add this correction.
[raw2rggb.sh.txt](https://github.com/FLIF-hub/FLIF/files/43668/raw2rggb.sh.txt)

**Comment by StephanBusch** [comment]
@psykauze Thank you for the .bat file.
Since my benchmarks compare more compressors on exctly the same input files, rotating would feel like cheating so I can test that script but I cannot publish that results because then the other compressors also need to be tested with rotated files. This would be different, of course, if Flif itself would detect CFA pattern and if it would rotate automatically.
My benchmark was created to see how far we can push lossless compression of RGGB and I really do appreciate the excellent work all of you do here. And I am very curious how your proposed MDL transform would work like and what further compression gains it will unlock.
Thank you very much for all your efforts

**Comment by psykauze** [comment]
I'm so sorry to read that :'(
The shell and *.bat files is for dev purpose, we know there is an issue about metadatas in the encoder. I consider this thing like a tweak, not a cheat

So there is some solutions I consider:
1) Working on rggb CFA detection internally but until we can't store the CFA Pattern in the flif file, it would be impossible to restore the original rggb file but we will have a great RGB file.
2) We can also rotate the file internally so the rggb output file can be decoded but rotated.
3) Developing the "channel sorting method" for YIQ (or MDL) transform will make the "CFA Pattern detection" useless (at least for "square" CFA Pattern)

**Comment by jonsneyers** [comment]
Storing metadata is trivial: just put it in a file and include it in the
ar, see `raw2flif` which extracts the exiv metadata and puts it in the
.flif file.
It would be cleaner if we use libraw and avoid shell scripts, and perhaps
let the decoder take rotation/CFA metadata into account directly.

**Comment by StephanBusch** [comment]
@psykauze I decided to run those tests with your batch because I want to help you developing FLIF by testing. I would also want to have a batch file that produces a .flif.
Instead of the linux-based ar you can use zlib or zstd.

Would it be better to design .flif with metadata as a compressed archive such as LibreOffice/MS Office or Mmiya-Leaf .eip files and store all metadata in a XML inside that archive?
Or would it be better to attach metadata directly to .flif as it is with all original raw files?

I have downloaded raw2rggb.bat and put exiftool.exe, dcraw.exe sed.exe (with libintl3.dll, regex2.dll and libiconv2.dll). Then tried to run raw2rggb.bat canon_eos_5d_mark_iii_05.cr2 canon.rggb but I got an error message: "(" cannot be syntactically processed at this position.

**Comment by psykauze** [comment]
OK, I've found the issue(s). I'm sorry, I've made some corrections but I haven't tested the bat file :( (Reason, I have not Windows at home). I'm working on it.

**Comment by jonsneyers** [comment]
Well, currently FLIF supports metadata as arbitrary files, so it could be compressed or plaintext XML, or whatever. The FLIF decoder simply ignores all metadata at the moment, but perhaps we should standardize _some_ metadata in order to allow `viewflif` to do something sensible with raw files. Information about rotation, CFA layout and channel multipliers would be needed (and sufficient?) to produce a reasonable non-interpolated, half-resolution image.

It would be nice if we can eventually losslessly convert DNG to FLIF and back without any shell/bat scripts, perhaps using `libraw`.

**Comment by StephanBusch** [comment]
@psykauze 
thank you very much

@jonsneyers 
yes, that would be my biggest wish - convert DNG to FLIF and then back to DNG again :))

@psykauze @jonsneyers 
Some men see things as they are and say "why?".
We dream of things that never were and say "Why not?"

**Comment by psykauze** [comment]
Sorry @StephanBusch, I've tried to debug the bat script but for some reason I've got random results (same input file, output file size change each try).

You could use the previous version of bat file. It was tested and it works.

**Comment by psykauze** [comment]
@StephanBusch @jonsneyers The bat script is now working. I've got some issues due to Batch scripting limitation.

[raw2rggb.txt](https://github.com/FLIF-hub/FLIF/files/45536/raw2rggb.txt)
[raw2rggb.bat.txt](https://github.com/FLIF-hub/FLIF/files/45548/raw2rggb.bat.txt)

**Comment by StephanBusch** [comment]
thank you for the update
I tried it with: raw2rggb.bat canon_eos_5d_mark_iii_05.cr2 c1.rggb

and no .rggb was created. The message was:
D:\TESTSETS\TEST_CAMERA_flif_rotate>IF "RGGBRGGBRGGBRGGB" == "RGGB" (SET Rotate=-t 0 )  ELSE IF "RGGBRGGBRGGBRGGB" == "GRBG" (SET Rotate=-t 270 )  ELSE IF "RGGBRGGBRGGBRGGB" == "BGGR" (SET Rotate=-t 180 )  ELSE IF "RGGBRGGBRGGBRGGB" == "GBRG" (SET Rotate=-t 90 )  ELSE (
ECHO This pattern is unknown: RGGBRGGBRGGBRGGB
 EXIT /B 1
)
This pattern is unknown: RGGBRGGBRGGBRGGB

D:\TESTSETS\TEST_CAMERA_flif_rotate>

on next image canon_eos_6d_14 the error message is:
SET PNMTYPE=P6
 GOTO GT_DECOMP
)  ELSE (
ECHO This raw color is unknown:
 EXIT /B 1
) )
This raw color is unknown:

**Comment by psykauze** [comment]
Please use this dcraw instead:
http://www.centrostudiprogressofotografico.it/en/dcraw/ (rename the exe as dcraw.exe)

Also, could you edit the bat and delete the first "REM" ? (Just the word, not the line)

**Comment by StephanBusch** [comment]
thank you.. using the provided DCRAW helped but not on the fujifilm_x_e1_20.raf.. the message here is:

D:\TESTSETS\TEST_CAMERA_flif_rotate>IF "GGRGGBGGBGGRBRGRBGGGBGGRGGRGGBRBGBRG" == "RGGB" (SET Rotate=-t 0 )  ELSE IF "GGRGGBGGBGGRBRGRBGGGBGGRGGRGGBRBGBRG" == "GRBG" (SET Rotate=-t 270 )  ELSE IF "GGRGGBGGBGGRBRGRBGGGBGGRGGRGGBRBGBRG" == "BGGR" (SET Rotate=-t 180 )  ELSE IF "GGRGGBGGBGGRBRGRBGGGBGGRGGRGGBRBGBRG" == "GBRG" (SET Rotate=-t 90 )  ELSE (
ECHO This pattern is unknown: GGRGGBGGBGGRBRGRBGGGBGGRGGRGGBRBGBRG
 EXIT /B 1
)
This pattern is unknown: GGRGGBGGBGGRBRGRBGGGBGGRGGRGGBRBGBRG

**Comment by psykauze** [comment]
Yeah. This is not an issue. This file can't be performed as a RGGB file in FLIF. The result will be worse than converting it in PGM mode.

**Comment by StephanBusch** [comment]
I get 11.098.920 bytes when input is .pgm
and 11.064.698 bytes if input is .rggb

**Comment by StephanBusch** [comment]
after trying to compress the first canon image, I get an error with FLIF:

===  flif -v -v -v -v -n c1.rggb c1.flif ===
  ____ _[_]____
 [__  | | |  __]   FLIF 0.1.9 [23 November 2015]
  |_  | |_|  _|    Free Lossless Image Format
    ]_|___|_[      (c) 2010-2015 J.Sneyers & P.Wuille, GNU GPL v3+

Loading input file: c1.rggb  RGGB file should be a PGM, like the output of "dcraw -E -4". Cannot read other types.
Could not read input file: c1.rggb

Execution time: 2.985 s

**Comment by psykauze** [comment]
I'm sorry but the script work as I expected. This file has a "weird" CFAPattern. FLIF can encode and decode it as a RGGB file without issues but if we try to decode it as a pam (or PNG) the result will be weird.
I prefer the script fails for this kind of file than not be sure of the result.

**Comment by StephanBusch** [comment]
you mean the fuji file in the above message?

**Comment by psykauze** [comment]
Yeah.

I will check the canon issue maybe it's just because your flif is not from the last commit.

**Comment by jonsneyers** [comment]
It isn't, as you can tell from the date. Renaming it to `.pnm` should do the trick, @StephanBusch 

@psykauze: I agree with not pretending that weird CFA patterns are actually RGGB. What do you think is the best approach to handling such files?

**Comment by StephanBusch** [comment]
if I use .pnm extension, only the PLC transform will be applied and the canon file is treated as 1 channel 16 bit: compression is also worse than it would be with .rggb extension

**Comment by jonsneyers** [comment]
Yes but if the file is not a PGM (P5) then it doesn't contain Bayered data anyway, but RGB pixels.

**Comment by StephanBusch** [comment]
I thought that raw2rggb.bat script creates RGGB rather than RGB pixels

**Comment by psykauze** [comment]
I found out why flif fails to decode the canon rggb file. It's because there is a 'space' caracter after P5 ("P5 \n" instead of "P5\n").

The new bat will not produce this 'space' 
[raw2rggb.bat.txt](https://github.com/FLIF-hub/FLIF/files/45694/raw2rggb.bat.txt)

**Comment by jonsneyers** [comment]
hm, I'm not sure about the PGM spec but if it allows that space to be there, then this is a bug in the rggb/pgm reader...

**Comment by StephanBusch** [comment]
The problem is still there - FLIF cannot read the input file.

**Comment by psykauze** [comment]
@jonsneyers The fuji x e1 CFA pattern contain 8 red pixels, 8 blue pixels and 20 green pixels. We can't debayerize this file in 4 planes with the same dimensions.
1) Easiest way: Encoding it only in greyscale mode
2) Not so easy but bitstream changing: Encoding each planes alone (like RGB mode) but with size different plane
3) Harder: Allowing flif to use more than 4 planes. For fuji x e1, 2 blue + 2 red + 5 green = 9 planes. YIQ and MDL transform can be applied.

@StephanBusch It seems the batch file add an extra non-printable character (0x0d) before the end of line. I think this is a flif issue not really a raw2rggb.bat issue.

[raw2rggb.bat.txt](https://github.com/FLIF-hub/FLIF/files/45755/raw2rggb.bat.txt)

**Comment by conradgaunt** [comment]
hello ,
  im not a c coder, im a pascal coder ..so i won't understand flif code easily .. but ill share my limited knowledge re: raw / how ive been dealing with it in my own adventures .. although i only came on here to ask a question ..

 first , a bit about me .. last march (2017) i spent a week throwing together a huffman coder , unhappy with 48bit cinema/ real world image compression options .. not sure what to expect...more as a learning exercise, mixed with frustration at being trapped inside 1980's image formats ...  after 2 days i had imagemagick pumping out RGB files , and im compressing (easy)...and more importantly ..5 days later ..decompressing!  (..got stoned after the success of compressing ....)  ..compression results weren't great .. but reasonable .. so then i tried pre-filtering the image with a dumb filter (no logic like paeth etc). at that point a 300tb image archive test showed that change made files 56% smaller than the openj2k encoder managed (on average / overall etc) ..flif was a bit better of course, at about 59% smaller ..but i didn't know that for another 3weeks ..literally (*). i have j2k files that come out smaller than flif btw, just a few. my system never beats flif , but stays quite close. annoyingly , the dumb filter / first attempt/ placeholder filter seems to be the best ive come up with .. which should be good , but it means since ive picked up the code after a year long break (got stoned again) ..ive now spend several weeks wasting my time testing alternatives without results .. but i do have 7 different reversible color transforms now , and they are ALL useful / get used (hint)

 during my initial experiments i also made a modified version to handle bayer patterns . i downloaded the blackmagic cinema camera sample files (all 11gig) , .. DCRAW can output a text file with meta data from the file read , i parse that , and get the RGGB order .. but ive only implemented the one BMCC uses out of laziness .
 anyway, thats enough about my unfeasibly sussessful attempts at compressing images losslessly

as for raw ..my advice for handing odd bayer patterns is to just decode them to full RGB . DCRAW can output 16bit linear bayer patterns , but odd patterns i just do a full decode , then throw away 2/3 of the values in any bayer configuration i want (or more precisely .. the same one BMCC uses..as ive implemented it already!)..reverse engineering the raw.. HOWEVER, what i intend to do in the future is count the number of original values in the red green and blue channels , and pick the bayer battern that maximizes the number of original values , OR has the widest dynamic range , with the latter trumping the former.  as im compressing sequences of large images , ill do this analysis once / or a few times per sequence of images (same when choosing the correct RCT),.. then assume that pattern for all images in the sequence (or test images from the start/ middle / end of sequences and try using any filter already chosen as optimal by one of those etc) , but its pretty quick analysis.  you can't cater of every proprietory pattern , unless perhaps you store metadata that can describe colors other than pure Red, Green, or Blue ( ie, by specify each bayer color component as a seperate set of R/G/B values in the metadata, yellow and purple filters can be used and colour adjusted .. but because im lazy , and manufacturers making sensors with purple and yellow filters can just f**k off / are perverts ..). the reality is , the physical colour filters used on the sensor will not be pure red green or blue anyway , and if you throw away the same amount of information that was never captured by the sensor taking the image in the first place , especially after its been fully decoded .. my tests show that reconstruction using simple bi-linear interolation never produces artifacts 

ive been focusing on debayering recent;y..wondering if adaptive debayering techniques can cause jitter in moving images. no artifacts appears in decoded redcode footage when simple bi-linear interpolation is used to debayer a "pseudo-reverse engineered raw file",..even after the image is downsampled to quarter resolution .. so im starting to think thats how red debayer too (no doubt because the sensor has a strong OLPF dropping optical resolution/ edge contrast making demosaicing trivial) . 
 ive been calling these pseudo raw images "documentary format" , as they save a LOT of space compared to the original 48bit TIFFs i generally get to play with . optionally , the pixels can be converted to 16bit half floats (a lossy transformation resulting in even smaller files , but identical dynamic range) .the main reason im trying to re-invent the wheel is because im not prepared to store more pixels than the cameras i use captured at the time,  long -term  .. not until storage is free 

as for compressing DNG files (or as i call them .. dung files ) ., i reencoded the blackmagic sample files , and compressed them using the correct bayer pattern , think i knocked off between 10-20%.  i used the bayer reversible colour transform described on the microsoft research site as the RCT option , and i  prefiltered the raw files before compressing by simply sweeping right to left across each image row , and subtracting the value 2 places to the left (filtering right to left during encoding, filtering left to right while decoding etc ) .. but then i also filter the filtered results vertically (from bottom to top, again, subtracting the value 2 rows above from the current value) . the results of pre comrpession filtering are signed values btw, bit depth arbitrary 

not sure if any of that is useful , but feel free to reach out to discuss my ramblings etc/ 

so ..i do have a couple of questions myself .. which is why i came on here  :
1) is flif ready for prime time yet , or still work in progress ?  .. btw, id like you to add more RCT's ;)
2) i couldn't get the flif encoder on the flif site to compress anything 48bit , so im using imagemagick to deal / make FLIF files.. are there any paramters for image magick that anyone knows of , such that i can do a brute force optimization using all available options ..and is it possible to encode bayer patterns with flif (i know it is ... but using imagemagick?) 

thanks in advance

k.r

**Comment by conradgaunt** [comment]
ps, final thought,

  many of the RCT's ive used in my compression system were designed by other people, and based around their understanding of how the human eye/ brain works etc .. however , lossless compression has nothing to do with how the eye works, or the brain recieves colour .. its just about reducing bytes ..but a lot of research papers i read start with lossy transforms yuv etc .. then elaborate them into lossless versions .. why!! i think zig zag patterns largely fall into that category . i believe the zig zag pattern is more useful in a lossy context, and is used for perceptual reasons , less useful in a lossless context
